### ISSUE TITLE
App using lots of CPU

### ISSUE THREAD

**Issue by tomwilkie** [weaveworks/scope#854]

**Comment by tomwilkie** [comment]
I think this can be closed now; its the probe using CPU.

**Comment by 2opremio** [comment]
While testing the 0.12 candidate ( d0d6cac) with the [ECS demo](http://www.weave.works/guides/service-discovery-and-load-balancing-with-weave-on-amazon-ecs-2/) I see that the app in one of the three hosts is consuming ~50% CPU with peaks of ~75% CPU. The other apps are running with less than 10% CPU consumption.

The app showing the problem is the only one whose UI I had opened in my browser

It doesn't happen systematically. I managed to reproduce it once yesterday and spent the rest of the day retrying without luck until I saw it happening today again.


Here are two CPU profiles:

 http://filebin.ca/2V8B4tkAgwWP/pprof.ec2-52-28-149-93.eu-central-1.compute.amazonaws.com4040.samples.cpu.001.pb.gz
http://filebin.ca/2V8Bmv9bJ5Xf/pprof.ec2-52-28-149-93.eu-central-1.compute.amazonaws.com4040.samples.cpu.002.pb.gz

And here's the png from one of them:

![profile](https://cloud.githubusercontent.com/assets/2362916/12676900/552abb52-c68d-11e5-913c-d0581a839984.png)


Which suggests it's the GC.

Here's the memory consumption profile: http://filebin.ca/2V8F07Fc8QZr/pprof.ec2-52-28-149-93.eu-central-1.compute.amazonaws.com4040.inuse_objects.inuse_space.001.pb.gz

![memory_profile](https://cloud.githubusercontent.com/assets/2362916/12677064/56ee3a44-c68e-11e5-914e-eb8c4a27026a.png)

Note: This is happening in a node whose Scope probe is experiencing a bad memory leak (#881), that the time I took the measurements, the Probe was consuming ~176MB, which may be related

**Comment by 2opremio** [comment]
This is the profile of the App (2ed39682ab05cd00b6c2d8d6127809ad7b77907c) at 70% CPU while monitoring the 5 probes in the scope.weave.works service: http://filebin.ca/2VkNcTTMcyob/pprof.localhost_4040.samples.cpu.001.pb.gz


![app](https://cloud.githubusercontent.com/assets/2362916/12801298/7d0ffef8-cad3-11e5-9b53-c5b8ffa0c0a1.png)

```
$ go tool pprof ~/home-vagrant/scope-tested pprof.localhost_4040.samples.cpu.001.pb.gz 
Entering interactive mode (type "help" for commands)
(pprof) list app.RegisterReportPostHandler.func1
Total: 18.57s
ROUTINE ======================== github.com/weaveworks/scope/app.RegisterReportPostHandler.func1 in /go/src/github.com/weaveworks/scope/app/router.go
         0     16.01s (flat, cum) 86.21% of Total
         .          .    122:                   rpt    report.Report
         .          .    123:                   reader = r.Body
         .          .    124:                   err    error
         .          .    125:           )
         .          .    126:           if strings.Contains(r.Header.Get("Content-Encoding"), "gzip") {
         .       40ms    127:                   reader, err = gzip.NewReader(r.Body)
         .          .    128:                   if err != nil {
         .          .    129:                           http.Error(w, err.Error(), http.StatusBadRequest)
         .          .    130:                           return
         .          .    131:                   }
         .          .    132:           }
         .          .    133:
         .          .    134:           decoder := gob.NewDecoder(reader).Decode
         .          .    135:           if strings.HasPrefix(r.Header.Get("Content-Type"), "application/json") {
         .          .    136:                   decoder = json.NewDecoder(reader).Decode
         .          .    137:           }
         .     15.97s    138:           if err := decoder(&rpt); err != nil {
         .          .    139:                   http.Error(w, err.Error(), http.StatusBadRequest)
         .          .    140:                   return
         .          .    141:           }
         .          .    142:           a.Add(rpt)
         .          .    143:           if len(rpt.Pod.Nodes) > 0 {
(pprof) 
```

Same behavior as above. Having a closer look at profile, the app spends most of the time

1. Decoding gobs, allocating memory and garbage collecting. 
2. Compiling gob decoders!

I am not super familiar with the App, but I guess we moved to gobs for efficiency reasons. However, I think they are optimized for continuous streamed communication and not individual REST calls. From http://blog.golang.org/gobs-of-data:

> The first time you send a given type, the gob package includes in the data stream a description of that type. In fact, what happens is that the encoder is used to encode, in the standard gob encoding format, an internal struct that describes the type and gives it a unique number
> After the type is described, it can be referenced by its type number.

Here's what I think is happening: the type information is included in every report, causing the gob decoder (and maybe the encoder too, I still have to check the probe profile) to be created (compiled) every single time an http request arrives.

So I would suggest to either:

1. Move the report endpoint from http into a websocket
2. Move away from gob into an encoding more efficient for http
3. Somehow cache the decoder compilation, but (understandably), `encoding/gob` doesn't seem to offer a way to do this. Maybe we can use do some trick with [Gobencoder/Gobdecoder](https://github.com/golang/go/blob/master/src/encoding/gob/decode.go#L1080)

If we take (2) maybe we could use a decoder which used Polls of Reports, also fixing the GC problem.

**Comment by 2opremio** [comment]
Maybe there's even a 4th, quick and hacky option: 

(4). Create a reader which concats all the http bodies for the decoder. But mixing this with gobs from different probes may be problematic or even make it useless.

**Comment by 2opremio** [comment]
Actually this seems to confirm my theory, I should probably had started there:

https://golang.org/pkg/encoding/gob/
> The implementation compiles a custom codec for each data type in the stream and is most efficient when a single Encoder is used to transmit a stream of values, amortizing the cost of compilation.

**Comment by tomwilkie** [comment]
I'd suggest trying Json first. Then websocket second. Your right, gob is a bad fit for this.

**Comment by 2opremio** [comment]
Using json ( #916 ) improves the App's situation considerably without impacting the probes. The app is now at 30% CPU with peaks of 40%. However, the decoding completely dominates the execution time:

[pprof.localhost:4040.samples.cpu.001.pb.gz](https://github.com/weaveworks/scope/files/117696/pprof.localhost.4040.samples.cpu.001.pb.gz)


![app_json](https://cloud.githubusercontent.com/assets/2362916/12820278/4ba5ac2c-cb56-11e5-96ba-9c0ad8a972c1.png)

Plus (and this is an important point I didn't mention) this all happens without even opening the UI in a browser. When doing this it gets peak of ~70% CPU.

I could try websockets+gob next but I think we should just aim at a better-suited codec instead of compromising on REST

**Comment by 2opremio** [comment]
Using ffjson without code generation ( see #916 ) helps a bit. The app is now down between 25% with peaks of 35% CPU.

![app_ffjson](https://cloud.githubusercontent.com/assets/2362916/12823242/daa33ed8-cb62-11e5-91b6-a90d46e5905a.png)

And it also seems to help when connecting the UI, when it gets at 60% CPU.

I guess this already starts to make Scope usable with the service.

**Comment by 2opremio** [comment]
Apart from optimizing the codecs, @bboreham made a fair point that we need to reduce the report sizes/transfer rates to the apps. I measured the size of the (uncompressed) gob reports and they are around 1MB each

```
<app> DEBU: 2016/02/11 11:09:34.664072 Decoded report with uncompressed size: 883188 bytes
<app> DEBU: 2016/02/11 11:09:38.115087 Decoded report with uncompressed size: 1025674 bytes
<app> DEBU: 2016/02/11 11:09:38.342998 Decoded report with uncompressed size: 935714 bytes
<app> DEBU: 2016/02/11 11:09:38.426127 Decoded report with uncompressed size: 1031151 bytes
<app> DEBU: 2016/02/11 11:09:38.815101 Decoded report with uncompressed size: 888347 bytes
<app> DEBU: 2016/02/11 11:09:41.632514 Decoded report with uncompressed size: 1039347 bytes
```

Probes send regular reports once every 3 seconds (and we also have asynchronous reports once every 10 seconds from the kubernetes and docker reporters) so the data processing requirements of the app get very quickly out of hand.

**Comment by 2opremio** [comment]
After the codec improvements here's the CPU profile of the app running in the service, with two probes attached and with its UI connected to a browser.

[pprof.localhost:4040.samples.cpu.001.pb.gz](https://github.com/weaveworks/scope/files/136696/pprof.localhost.4040.samples.cpu.001.pb.gz)


![app](https://cloud.githubusercontent.com/assets/2362916/13153285/28d90e80-d66b-11e5-9b9d-3c5d63d7cc93.png)

The main bottlenecks are similar to the ones in the probe (https://github.com/weaveworks/scope/issues/812#issuecomment-185840463):
* Copying/Merging which causes a lot of garbage
* Decoding ( #985 should improve this)

**Comment by 2opremio** [comment]
Here's the `--alloc_objects` heap profile for the app (d7442788548c53971aaccdb681a274db31ba57a3) running in the service (connected to 4 probes):

[pprof.localhost:4040.alloc_objects.alloc_space.001.pb.gz](https://github.com/weaveworks/scope/files/138491/pprof.localhost.4040.alloc_objects.alloc_space.001.pb.gz)


![app_obj_allocs](https://cloud.githubusercontent.com/assets/2362916/13185001/e68bb7fe-d736-11e5-8032-1ca1586de40b.png)

**Comment by 2opremio** [comment]
After #1000 the codec is not generating almost any garbage compared to the immutable datastructures, copies and merge operations. The GC is clearly the bottleneck. 

Note how `runtime.scanobject` accounts for almost 40% of the CPU time and how `ps.Cons`+`ps.Set` generate almost 30% of the garbage.

CPU profile:

[pprof.localhost:4043.samples.cpu.004.pb.gz](https://github.com/weaveworks/scope/files/140908/pprof.localhost.4043.samples.cpu.004.pb.gz)


![app](https://cloud.githubusercontent.com/assets/2362916/13223824/c58e9bea-d97d-11e5-9e7e-f94a9081a7b2.png)


Object allocation profile:

[pprof.localhost:4043.alloc_objects.alloc_space.001.pb.gz](https://github.com/weaveworks/scope/files/140910/pprof.localhost.4043.alloc_objects.alloc_space.001.pb.gz)

![app_object_alloc](https://cloud.githubusercontent.com/assets/2362916/13223854/ebbaad54-d97d-11e5-91ea-ed6a369a10d2.png)


Note how cons generates 20% of the garbage.

**Comment by 2opremio** [comment]
Closing in favor of #1010
